File: 3. Behavioral Design Patterns/Mediator/C#/MediatorTemplate.cs
// Mediator Pattern – C# Template

using System.Collections.Generic;

namespace MediatorTemplate
{
    interface IMediator { void Notify(object sender, string eventType); }

    abstract class Colleague
    {
        protected IMediator mediator;
        public Colleague(IMediator mediator) { this.mediator = mediator; }
        public abstract void DoSomething();
    }

    class ConcreteColleagueA : Colleague
    {
        public ConcreteColleagueA(IMediator mediator) : base(mediator) {}
        public override void DoSomething() { mediator.Notify(this, "EventA"); }
    }

    class ConcreteColleagueB : Colleague
    {
        public ConcreteColleagueB(IMediator mediator) : base(mediator) {}
        public override void DoSomething() { mediator.Notify(this, "EventB"); }
    }

    class ConcreteMediator : IMediator
    {
        private readonly List<Colleague> colleagues = new();
        public void AddColleague(Colleague colleague) { colleagues.Add(colleague); }
        public void Notify(object sender, string eventType) { /* handle event */ }
    }

    class Program
    {
        static void Main()
        {
            var mediator = new ConcreteMediator();
            var colleagueA = new ConcreteColleagueA(mediator);
            var colleagueB = new ConcreteColleagueB(mediator);
            mediator.AddColleague(colleagueA);
            mediator.AddColleague(colleagueB);
            colleagueA.DoSomething();
        }
    }
}

File: 3. Behavioral Design Patterns/Mediator/Java/MediatorTemplate.java
// Mediator Pattern – Java Template

import java.util.*;

interface Mediator { void notify(Object sender, String eventType); }

abstract class Colleague {
    protected Mediator mediator;
    public Colleague(Mediator mediator) { this.mediator = mediator; }
    public abstract void doSomething();
}

class ConcreteColleagueA extends Colleague {
    public ConcreteColleagueA(Mediator mediator) { super(mediator); }
    public void doSomething() { mediator.notify(this, "EventA"); }
}

class ConcreteColleagueB extends Colleague {
    public ConcreteColleagueB(Mediator mediator) { super(mediator); }
    public void doSomething() { mediator.notify(this, "EventB"); }
}

class ConcreteMediator implements Mediator {
    private final List<Colleague> colleagues = new ArrayList<>();
    public void addColleague(Colleague colleague) { colleagues.add(colleague); }
    public void notify(Object sender, String eventType) { /* handle event */ }
}

public class MediatorTemplate {
    public static void main(String[] args) {
        ConcreteMediator mediator = new ConcreteMediator();
        Colleague colleagueA = new ConcreteColleagueA(mediator);
        Colleague colleagueB = new ConcreteColleagueB(mediator);
        mediator.addColleague(colleagueA);
        mediator.addColleague(colleagueB);
        colleagueA.doSomething();
    }
}
